### PR TITLE
periph/gpio: add gpio_init_low() / gpio_init_high() helper functions

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description
Provide a way to init a GPIO as output with a set level.
Some platforms can to this on init already and will always just set
the GPIO to LOW which can be undesireable.

For all other platforms, provide a fall-back implementation.

The functions are also convenient when de-initializing a previously
otherwise configured Pin, e.g. when disconnecting the power on an external
sensor through a transistor and wanting to avoid current leaking through a
high idle UART TX line.

### Testing procedure

The code is trivial, but you can check check that a pin is really HIGH after calling `gpio_init_high(GPIO_PIN(0,1));` 

### Issues/PRs references
Provides a clean solution for #12380

### CODE REVIEW

**Comment by basilfx** [comment]
I created a feature branch where I changed the behaviour for EFM32 w.r.t. this issue. If that works in #12380 then it should fix this issue for EFM32.

**Comment by benpicco** [comment]
@basilfx want to give this a review?

**Comment by basilfx** [comment]
@benpicco I gave this PR a review and I approve it partially.

Since this extends the GPIO API, I don' feel comfortable to ACK for that part. The change does make sense, and it is only a helper function that doesn't affect the rest.

Also, I think it needs at least a test that hits these two new functions.

**Comment by basilfx** [comment]
@benpicco I think I got you covered for a test case: https://github.com/basilfx/RIOT/commit/173c05f76f1764e01bec38fc70cff00a63caa1c9.

Feel free to cherry-pick that one.

**Comment by benemorius** [comment]
I feel strongly that the API should include a way to set the driven level of a GPIO in the same call as setting the mode to output, so I'm in favor of this PR.

In fact I even feel that the API *shouldn't* include a way to set the mode to output *without* also defining the driven level in the same call, but of course that's outside the scope of this PR.

At a glance, the only thought I had about the specifics of the API is that perhaps it's preferable to have high/low passed as a parameter rather than having two separate functions to call. Then again, perhaps it isn't. Maybe both. Both would be consistent with the fact that we have both `gpio_write()` and `gpio_set()`/`gpio_clear()`.

File: drivers/periph_common/gpio.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * @brief       Fallback Implementations for platform specific GPIO routines.
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+
+#include "periph/gpio.h"
+
+#ifndef GPIO_HAVE_INIT_LEVEL
```
Comment (maribu) [inline drivers/periph_common/gpio.c:23]:
```suggestion
#if !defined(GPIO_HAVE_INIT_LEVEL) || defined(DOXYGEN)
```

File: tests/periph_gpio/main.c [hunk 2]
```diff
@@ -55,11 +57,48 @@ static int init_pin(int argc, char **argv, gpio_mode_t mode)
     return 0;
 }
 
+static int init_pin_hl(int argc, char **argv, bool high)
+{
+    int po, pi, r;
```
Comment (maribu) [inline tests/periph_gpio/main.c:62]:
```suggestion
    int port, pin, r;
```

Only three letters more, but much more readable ;-)

File: drivers/periph_common/gpio.c (hunk 1 above)
Comment (benpicco) [reply drivers/periph_common/gpio.c:23]: Does Doxygen care about C files?

File: tests/periph_gpio/main.c (hunk 2 above)
Comment (benpicco) [reply tests/periph_gpio/main.c]: I also changed `init_pin` while I'm at it.

Comment (maribu) [reply tests/periph_gpio/main.c]: No. I somehow assumed this to be `static inline` in the header. I was scrolling to fast over the diff :-/

File: drivers/periph_common/gpio.c (hunk 1 above)
Comment (maribu) [reply drivers/periph_common/gpio.c:23]: No. I somehow assumed this to be `static inline` in the header. I was scrolling to fast over the diff :-/

**Comment by maribu** [comment]
> That's probably a bug in that implementation. gpio_init() should not touch the currently configured level.

Sadly, this is not always easily possible. E.g. for the STM32F446 the pin used for UART is high, when it is not currently configured. If you initialize it (without touching the data register), it becomes low.

To properly implement it, I assume one would have to prepare a look up table with which GPIOs should be high upon init and which should be low. I think these functions are less pain.

**Comment by maribu** [comment]
It might additionally make sense to still have a `gpio_mode_t mode` parameter. E.g. if I would have to implement I2C in software, I would like to use `gpio_init_high(scl_pin, GPIO_OD_PU)`.

File: drivers/periph_common/gpio.c [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    return gpio_init(pin, GPIO_OUT);
+}
+
+int gpio_init_high(gpio_t pin)
+{
+    gpio_set(pin);
+    return gpio_init(pin, GPIO_OUT);
+}
+
+#endif /* GPIO_HAVE_INIT_LEVEL */
```
Comment (maribu) [inline drivers/periph_common/gpio.c:36]:
```suggestion
#else
typedef int dont_be_pedantic;
#endif /* GPIO_HAVE_INIT_LEVEL */
```

Formally, an empty compilation unit is not allowed. A `typedef` is formally considered as a statement. Thus, a bogus `typedef` can make sure that this compilation unit is always valid C.

File: drivers/periph_common/gpio.c [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "periph/gpio.h"
+
+#ifndef GPIO_HAVE_INIT_LEVEL
+
+int gpio_init_low(gpio_t pin)
+{
+    gpio_clear(pin);
+    return gpio_init(pin, GPIO_OUT);
+}
```
Comment (maribu) [inline drivers/periph_common/gpio.c:28]: I'm not sure if the fallback implementation should rely on this to work. I see to alternatives:

1. Introduce a new feature. This implementation could use `weak` symbols, so that platforms actually support `gpio_clear()` or `gpio_set()` being called prior to initialization, this can be explicitly enabled and no code needs to be provided.
2. Initialize the pin first and clear it afterwards. That works reliable in for every platform, only that the quality of the experience of `gpio_init_low()` will not be superb on some platforms

Comment (benpicco) [reply drivers/periph_common/gpio.c]: I was under the impression that 

```C
gpio_set()
gpio_init()
```

was the canonical way to ensure a GPIO is set in RIOT.
This is used in a few places like that - so this is a direct replacement.

If the platform does not support this pattern, it should implement `gpio_init_low()`/`gpio_init_high()`.
Whether we use weak functions or a define doesn't make a difference, I tried to mirror `PERIPH_TIMER_PROVIDES_SET`.

Comment (maribu) [reply drivers/periph_common/gpio.c]: My point was not that I prefer the weak symbol over the define. I would like to have two knobs to fiddle with, one controlling whether `gpio_init_low()` / `gpio_init_high()` is supported at all and one to provide a custom implementation. So in total there should be three possible states:

1. The platform does not support `gpio_init_low()` / `gpio_init_high()` at all
2. The platform does support it using the common implementation (e.g. `gpio_set(); gpio_init()`)
3. The platform does support it using a platform specific implementation

Or alternatively, the fallback implementation should first initialize the pin and then sets/clears it. That is guaranteed to work, but suboptimal.

> was the canonical way to ensure a GPIO is set on init in RIOT.

IMO, this is not true, as the doc doesn't state this. But the GPIO doc is pretty poor anyway. Maybe you can fix this (at least partially) in this PR and just add a precondition (via `@pre`) that `gpio_set()` and `gpio_clear()` must not be called for pins that have not been initialized using one of the `gpio_init*()` functions to one of the output modes. That way `gpio_init_high()` and `gpio_init_low()` would become the canonical way.

(Other issues are IMO what the return value of `gpio_read()` is, if the pin is not configured as input (with or without pulls). E.g. for `GPIO_OUT`, I would intuitively expect it to return `!=0` after `gpio_set()` and `0` after `gpio_clear()`. But for `GPIO_OD_PU` I would expect `gpio_read()` after `gpio_set()` to potentially still return `0`, if the line is externally pulled low. But none of this is documented.)

File: drivers/periph_common/gpio.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+__attribute__((weak))
+int gpio_init_high(gpio_t pin, gpio_mode_t mode)
+{
+    assert(mode != GPIO_IN);
+    assert(mode != GPIO_IN_PD);
+    assert(mode != GPIO_IN_PU);
+
+    if (gpio_init(pin, mode)) {
+        return -1;
```
Comment (benpicco) [inline drivers/periph_common/gpio.c:46]: I'm very tempted to just `assert(0)` here - IMHO `gpio_init()` doesn't need a return value, it either works or it's a configuration problem.
Nothing a run-time check could hanlde.

Comment (maribu) [reply drivers/periph_common/gpio.c:46]: I think a PR to make `gpio_init()` return `void` has good chances of getting merged (after the feature freeze).

**Comment by fjmolinas** [comment]
ping @maribu @benpicco!

File: tests/periph_gpio/main.c [hunk 6]
```diff
@@ -272,6 +311,8 @@ static int bench(int argc, char **argv)
 
 static const shell_command_t shell_commands[] = {
     { "init_out", "init as output (push-pull mode)", init_out },
+    { "init_out_h", "init as high-output (push-pull mode)", init_out_h },
+    { "init_out_l", "init as low-output (push-pull mode)", init_out_l },
```
Comment (maribu) [inline tests/periph_gpio/main.c:315]: Can you rephrase the "high-output" / "low-output"? To me, this sounds like a special output mode.

Comment (benpicco) [reply tests/periph_gpio/main.c]: Ok, reworded.

File: tests/periph_gpio/main.c [hunk 7]
```diff
@@ -272,6 +311,8 @@ static int bench(int argc, char **argv)
 
 static const shell_command_t shell_commands[] = {
     { "init_out", "init as output (push-pull mode)", init_out },
+    { "init_out_h", "init as output high (push-pull mode)", init_out_h },
+    { "init_out_l", "init as output low (push-pull mode)", init_out_l },
```
Comment (maribu) [inline tests/periph_gpio/main.c:315]:
```suggestion
    { "init_out_h", "init as output (push-pull mode, high after init)", init_out_h },
    { "init_out_l", "init as output (push-pull mode, low after init)", init_out_l },
```

**Comment by kaspar030** [comment]
wouldn't it make more sense to add this to the gpio_init modes we already have?

**Comment by maribu** [comment]
> wouldn't it make more sense to add this to the gpio_init modes we already have?

If I understand correctly, you mean extending the enum to additionally contain `GPIO_OUTPUT_HIGH` and `GPIO_OUTPUT_LOW`.

I agree that from an API point of view this would be most consistent. However, we would have to touch a few GPIO implementations for that.

**Comment by kaspar030** [comment]
> I agree that from an API point of view this would be most consistent. However, we would have to touch a few GPIO implementations for that.

Is it that bad?

File: drivers/include/periph/gpio.h [hunk 8]
```diff
@@ -224,27 +249,39 @@ int gpio_read(gpio_t pin);
 /**
  * @brief   Set the given pin to HIGH
  *
+ * @pre     The given pin has been configured as Output using
```
Comment (kaspar030) [inline drivers/include/periph/gpio.h:252]: these seem now incomplete. I can set a pin to "high" that has been initialized using `gpio_init_low/high()`

**Comment by kaspar030** [comment]
> Most of them accomplish it by calling `gpio_write()` prior to ever calling `gpio_init()` which is an undocumented feature common to most platforms except notably `efm32` as referenced in OP.

This actually goes against documentation, periph "handles" are only allowed to be used after they have been init()ed correctly. That's our shortcut to not have to check the validity of e.g., a gpio_t parameter, on every set or clear.

File: drivers/include/periph/gpio.h (hunk 8 above)
Comment (benpicco) [reply drivers/include/periph/gpio.h:252]: Well of course :wink:

**Comment by fjmolinas** [comment]
> This actually goes against documentation, periph "handles" are only allowed to be used after they have been init()ed correctly. That's our shortcut to not have to check the validity of e.g., a gpio_t parameter, on every set or clear.


@kaspar030 should we add the contested tag to this one then?

**Comment by kaspar030** [comment]
> @kaspar030 should we add the contested tag to this one then?

no, that was just for the workaround mentioned (first `gpio_set()`, then gpio_init()`).

I'm not happy with adding and calling these `gpio_init_low/high()`, as this is for configuring output. That's somewhat confusing, IMO. We'd uglify the API so we don't have to touch the implementations.

Would an alternative be wrapping the actual implementations?
Like, renaming the actual per-cpu gpio_init() to gpio_init_sth(), and then using this PR's gpio_init_high/low as a wrapper, but changed to use additional enum values?

**Comment by benemorius** [comment]
> I'm not happy with adding and calling these `gpio_init_low/high()`, as this is for configuring output. That's somewhat confusing, IMO.

I'm trying but I'm struggling to understand which aspect you object to. Are you saying you'd be happier with `gpio_init(..., GPIO_OUT_HIGH)`?

As a hardware designer I object to an API that permits telling the hardware to become an output without simultaneously telling it which level to output. It's the very definition of undefined behavior.

Perhaps we would be better off with separate functions to init GPIOs as input versus output? After all from a hardware perspective it is fundamentally different. Inputs need pull resistors specified whereas outputs need the driven level specified.

**Comment by kaspar030** [comment]
> I'm trying but I'm struggling to understand which aspect you object to. Are you saying you'd be happier with `gpio_init(..., GPIO_OUT_HIGH)`?

yes, I mean only from an API perspective.

> As a hardware designer I object to an API that permits telling the hardware to become an output without simultaneously telling it which level to output. It's the very definition of undefined behavior.

:+1:

GPIO_OUT could be an alias for GPIO_OUT_LOW. That shouldn't be too hard to fix / confirm in the existing implementations?

**Comment by maribu** [comment]
What was missing here to get this in?

**Comment by fjmolinas** [comment]
ping @maribu @benpicco @kaspar030

**Comment by kaspar030** [comment]
> What was missing here to get this in?

I was objecting the API clutter:

- `gpio_init(..., mode)` initializing input, output *not touching state*
- `gpio_init_low(...) initializing output, setting to low
- `gpio_init_high(...) initializing output, setting to high

vs.

- `gpio_init(..., mode)` initializing input, output *not touching state*
- `gpio_init(..., GPIO_OUT)` initializing input, output *not touching state*
- `gpio_init(...., GPIO_OUT_LOW) initializing output, setting to low
- `gpio_init(..., GPIO_OUT_HIGH) initializing output, setting to high

`gpio_init_low()` kinda implies the initializiation of an output, but just kinda.

`gpio_init_out_low()` and `gpio_init_out_high()` would be a weird compromise, but why no `gpio_init_in/out/in_pd/out_pd/...()`?
Either the API is using a `mode` parameter, *or* distinct functions. Both are just inconsistend.
And I'd object changing towards non-overloaded init(). Too much change for too little benefit.

**Comment by maribu** [comment]
Note that `gpio_ll` would also scratch this itch and others ;-)
